addons/mail/static/src/models/thread_cache/thread_cache.js

```diff
@@ -94,18 +94,88 @@ function factory(dependencies) {
         // Private
         //----------------------------------------------------------------------
 
+        /**
+         * @private
+         * @returns {mail.message[]}
+         */
+        _computeFetchedMessages() {
+            if (!this.thread) {
+                return [];
+            }
+            const cmd = [];
+            for (const message of this.fetchedMessages) {
+                if (!this.thread.messages.includes(message)) {
+                    cmd.push(['unlink', message]);
+                }
+            }
+            return cmd;
```

Comment:
Build array and just do one unlink at the end?

Could even use unlink-filter with the condition as filter.